### PR TITLE
MODINVSTOR-361: Populate effective call number type id on item.

### PR DESCRIPTION
**Pull request by bohdan-suprun**
https://issues.folio.org/browse/MODINVSTOR-361: Store effective call number type on item record

* Store effective call number type id on item record.
* Update migration script.
* Update API version.
* Refactor existing item call number related test cases and create parametrized tests instead.

### CODE REVIEW

File: src/main/resources/templates/db_scripts/populateEffectiveCallNumberComponentsForExistingItems.sql [hunk 1]
```diff
@@ -5,7 +5,8 @@ SET jsonb = jsonb_set(
   jsonb_build_object(
     'callNumber', COALESCE(it.jsonb->'itemLevelCallNumber', hr.jsonb->'callNumber'),
     'prefix', COALESCE(it.jsonb->'itemLevelCallNumberPrefix', hr.jsonb->'callNumberPrefix'),
-    'suffix', COALESCE(it.jsonb->'itemLevelCallNumberSuffix', hr.jsonb->'callNumberSuffix')
+    'suffix', COALESCE(it.jsonb->'itemLevelCallNumberSuffix', hr.jsonb->'callNumberSuffix'),
+    'typeId', COALESCE(it.jsonb->'itemLevelCallNumberTypeId', hr.jsonb->'callNumberTypeId')
```
Comment (marcjohnson-kint) [inline src/main/resources/templates/db_scripts/populateEffectiveCallNumberComponentsForExistingItems.sql]: Can we do this here, because this script needs to be used for different versions of the module?

File: src/main/resources/templates/db_scripts/schema.json [hunk 2]
```diff
@@ -698,7 +698,7 @@
     {
       "run": "after",
       "snippetPath": "populateEffectiveCallNumberComponentsForExistingItems.sql",
-      "fromModuleVersion": "17.1.0"
+      "fromModuleVersion": "18.2.0"
```
Comment (marcjohnson-kint) [inline src/main/resources/templates/db_scripts/schema.json]: Hmm. How does this work, does this mean that this definition will only work for version 18.2.0?

File: src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+@RunWith(JUnitParamsRunner.class)
+public class ItemEffectiveCallNumberComponentsTest extends TestBaseWithInventoryUtil {
+  private static final String HOLDINGS_CALL_NUMBER_TYPE = UUID.randomUUID().toString();
+  private static final String HOLDINGS_CALL_NUMBER_TYPE_SECOND = UUID.randomUUID().toString();
+  private static final String ITEM_LEVEL_CALL_NUMBER_TYPE = UUID.randomUUID().toString();
+  private static final String ITEM_LEVEL_CALL_NUMBER_TYPE_SECOND = UUID.randomUUID().toString();
+
+  @BeforeClass
+  public static void createCallNumberTypes() throws Exception {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java]: What happens if these records already exist?

File: src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+  private List<Object[]> createPropertiesParams() {
+    List<Object[]> testCases = new ArrayList<>();
+
+    testCases.addAll(createPropertyTestCase("callNumber", "callNumber"));
+    testCases.addAll(createPropertyTestCase("callNumberSuffix", "suffix"));
+    testCases.addAll(createPropertyTestCase("callNumberPrefix", "prefix"));
+    testCases.addAll(createPropertyTestCase(
+      "callNumberTypeId", HOLDINGS_CALL_NUMBER_TYPE,
+      ITEM_LEVEL_CALL_NUMBER_TYPE, "typeId"
+    ));
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java]: Why has this line been split sooner than the line length limit?

File: src/test/java/org/folio/rest/api/ItemEffectiveCallNumberTest.java [hunk 5]
```diff
@@ -10,22 +13,29 @@
...
 
+import com.fasterxml.jackson.databind.ObjectMapper;
+
 import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
 import junitparams.JUnitParamsRunner;
+import junitparams.Parameters;
 
 @RunWith(JUnitParamsRunner.class)
 public class ItemEffectiveCallNumberTest extends TestBaseWithInventoryUtil {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/ItemEffectiveCallNumberTest.java]: What is the difference between this class and the `ItemEffectiveCallNumberComponentsTest`?

File: src/main/resources/templates/db_scripts/populateEffectiveCallNumberComponentsForExistingItems.sql (hunk 1 above)
Comment (bohdan-suprun) [reply src/main/resources/templates/db_scripts/populateEffectiveCallNumberComponentsForExistingItems.sql]: It is not so clear from the documentation. I've asked @TarasSpashchenko, it seems that it is fine to extend this migration file and change  the `fromModuleVersion`.

Am I correct understand that the inventory-storage module has already been released? And new code, merged to the master, won't affect the release?

File: src/main/resources/templates/db_scripts/schema.json (hunk 2 above)
Comment (bohdan-suprun) [reply src/main/resources/templates/db_scripts/schema.json]: Yes, for version >= `18.2.0`  I think.

File: src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java (hunk 3 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java]: I don't think it is possible, because the record will be created only once, and will have a random ID.

File: src/test/java/org/folio/rest/api/ItemEffectiveCallNumberTest.java (hunk 5 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/rest/api/ItemEffectiveCallNumberTest.java]: The `ItemEffectiveCallNumberTest ` it is test for the migration script, while `ItemEffectiveCallNumberComponentsTest` it is test for the update logic itself.

**Comment by bohdan-suprun** [comment]
> I find the parameterised tests quite hard to follow, so I think it would be useful for someone else more familiar with this style of testing to review this change

@marcjohnson-kint, Do you think I should replace them with the "usual", not parametrized, tests? Because my main goal was to reduce test duplications. We had ~3-4 test cases per each call number property which were pretty-much the same.

**Comment by marcjohnson-kint** [comment]
> > I find the parameterised tests quite hard to follow, so I think it would be useful for someone else more familiar with this style of testing to review this change
> 
> @marcjohnson-kint, Do you think I should replace them with the "usual", not parametrized, tests? Because my main goal was to reduce test duplications. We had ~3-4 test cases per each call number property which were pretty-much the same.

No, I think other folks have used parameterised tests. I personally find them indirect and hard to read, that is probably more my challenge than the tests.

File: src/test/java/org/folio/rest/api/ItemEffectiveCallNumberTest.java (hunk 5 above)
Comment (marcjohnson-kint) [reply src/test/java/org/folio/rest/api/ItemEffectiveCallNumberTest.java]: Could we rename `ItemEffectiveCallNumberTest` to something that reflect that it is for the migration script, e.g. `ItemEffectiveCallNumberDataUpgradeTest`?

File: src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java (hunk 3 above)
Comment (marcjohnson-kint) [reply src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java]: Are there any constraints on the name of the call number type?

My concern is that these records aren't cleaned up between tests.

File: src/main/resources/templates/db_scripts/populateEffectiveCallNumberComponentsForExistingItems.sql (hunk 1 above)
Comment (marcjohnson-kint) [reply src/main/resources/templates/db_scripts/populateEffectiveCallNumberComponentsForExistingItems.sql]: Yes, mod-inventory-storage has been released.

I think it is potentially confusing to alter existing scripts and change when they are executed.

Thanks for asking for clarification from folks.

File: src/main/java/org/folio/rest/support/EffectiveCallNumberComponentsUtil.java [hunk 6]
```diff
@@ -36,9 +39,17 @@ private static EffectiveCallNumberComponents updateComponents(
       updatedCallNumberSuffix = holdings.getCallNumberSuffix();
     }
 
+    String updatedCallNumberTypeId = StringUtils.firstNonBlank(
+      item.getItemLevelCallNumberTypeId(),
+      Optional.ofNullable(holdings).map(HoldingsRecord::getCallNumberTypeId)
+        .orElse(null)
```
Comment (mreno-EBSCO) [inline src/main/java/org/folio/rest/support/EffectiveCallNumberComponentsUtil.java]: I think we may be abusing `Optional`, I know I am guilty of it as well and I need to stop, but this is better suited for a ternary operator. `Optional` is really meant for return types as described here (from an Oracle blog):
> The purpose of `Optional` is not to replace every single null reference in your codebase but rather to help design better APIs in which—just by reading the signature of a method—users can tell whether to expect an optional value.
```suggestion
        holdings != null ? holdings.getCallNumberTypeId() : null
```

File: src/main/resources/templates/db_scripts/populateEffectiveCallNumberComponentsForExistingItems.sql (hunk 1 above)
Comment (bohdan-suprun) [reply src/main/resources/templates/db_scripts/populateEffectiveCallNumberComponentsForExistingItems.sql]: https://github.com/folio-org/raml-module-builder/blob/master/DB-schema-migration.md:

It seems that the only case when the script won't be executed is when `current module version` > `fromModuleVersion`.

For example: 
current module version is: 18.2.0.
fromModuleVersion is: 17.1.0.

In this case the script/table creation/etc. won't be executed because the RMB assumes that the change has already been applied before.

But for release example:
current module version (released) is: 17.1.0.
fromModuleVersion is: 18.2.0.

The migration script should be applied.

@TarasSpashchenko, @skoczko, @julianladisch am I correct understand?

File: src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java (hunk 3 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java]: Yes, we have a UK for the name field. I'm specifying a test-specific name, I don't think that we have the same name. Alternatively, I can re-use existing types, from reference data. Does it sound better?

File: src/test/java/org/folio/rest/support/builders/HoldingRequestBuilder.java [hunk 7]
```diff
@@ -38,7 +40,8 @@ private HoldingRequestBuilder(
     String callNumber,
     String callNumberPrefix,
     String callNumberSuffix,
-    String hrid) {
+    String hrid,
+    String callNumberTypeId) {
```
Comment (mreno-EBSCO) [inline src/test/java/org/folio/rest/support/builders/HoldingRequestBuilder.java]: This is just aesthetic, but I think we should keep the `callNumber*` components together in the argument list, i.e. swap `hrid` and `callNumberTypeId`.

File: src/main/resources/templates/db_scripts/populateEffectiveCallNumberComponentsForExistingItems.sql (hunk 1 above)
Comment (julianladisch) [reply src/main/resources/templates/db_scripts/populateEffectiveCallNumberComponentsForExistingItems.sql]: Yes.

File: src/main/resources/templates/db_scripts/schema.json (hunk 2 above)
Comment (julianladisch) [reply src/main/resources/templates/db_scripts/schema.json]: Yes, the populate script will be executed when upgrading from an old version that is `< 18.2.0`. For example when upgrading from 17.0.0 to 19.0.0. But not when upgrading from 18.2.0 to 19.0.0.

File: src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java (hunk 3 above)
Comment (marcjohnson-kint) [reply src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java]: I’m not suggesting reusing records from the reference data. I’d prefer the tests didn’t do that because I think they should be independent of it.

My concern is that these records don’t appear to be removed. Are they removed elsewhere in a general cleanup?

Comment (bohdan-suprun) [reply src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java]: I don't think, I can remove it in `@AfterClass` method.

Comment (marcjohnson-kint) [reply src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java]: I typically suggest cleaning up records before tests run rather than after, in order to aid diagnostics if tests fail.

File: ramls/item.json [hunk 8]
```diff
@@ -69,6 +69,12 @@
           "type": "string",
           "description": "Effective Call Number Suffix is the suffix of the identifier assigned to an item or its holding and associated with the item.",
           "readonly": true
+        },
+        "typeId": {
+          "type": "string",
+          "description": "Effective Call Number Type Id is the call number type id assigned to the item or associated holding.",
```
Comment (julianladisch) [inline ramls/item.json]: Can the description be clear which id is chosen if both item and holding have such an id?
For example "is the call number type id of the item, if available, otherwise that of the holding."

**Comment by julianladisch** [comment]
The tests in ItemEffectiveCallNumberComponentsTest.java are difficult to read because one cannot see which values gets passed into the test methods.
This is the same set of test values but visible from the source code: https://github.com/folio-org/mod-inventory-storage/blob/0b879ea9720bc86ba2222450b3058edd3cbfdd92/src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java

**Comment by bohdan-suprun** [comment]
> The tests in ItemEffectiveCallNumberComponentsTest.java are difficult to read because one cannot see which values gets passed into the test methods.
> This is the same set of test values but visible from the source code: https://github.com/folio-org/mod-inventory-storage/blob/0b879ea9720bc86ba2222450b3058edd3cbfdd92/src/test/java/org/folio/rest/api/ItemEffectiveCallNumberComponentsTest.java

Yo're right, but another side is that params string becomes too long for update case, and it will reduce readability.

Let me try to refactor them.

**Comment by bohdan-suprun** [comment]
@julianladisch, I've refactored tests, they should be more readable now, please take a look.

File: src/test/java/org/folio/rest/api/TestBaseWithInventoryUtil.java [hunk 9]
```diff
@@ -42,12 +44,12 @@
...
-  protected static final String ANNEX_LIBRARY_LOCATION = "Annex Library";
-  protected static final String ONLINE_LOCATION = "Online";
-  protected static final String THIRD_FLOOR_LOCATION = "Third Floor";
-  protected static final String FOURTH_FLOOR_LOCATION = "Fourth Floor";
+  public static final String MAIN_LIBRARY_LOCATION = "Main Library";
+  public static final String SECOND_FLOOR_LOCATION = "Second Floor";
+  public static final String ANNEX_LIBRARY_LOCATION = "Annex Library";
+  public static final String ONLINE_LOCATION = "Online";
+  public static final String THIRD_FLOOR_LOCATION = "Third Floor";
+  public static final String FOURTH_FLOOR_LOCATION = "Fourth Floor";
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/api/TestBaseWithInventoryUtil.java]: Why have locations needed to be made public for a call number related test?

File: src/test/java/org/folio/rest/support/http/ResourceClient.java [hunk 10]
```diff
@@ -160,22 +170,26 @@ public Response getById(UUID id)
...
+  public void delete(UUID id) throws MalformedURLException, InterruptedException,
+    ExecutionException, TimeoutException {
+
+    Response response = deleteIfPresent(id != null ? id.toString() : null);
 
     assertThat(String.format(
       "Failed to delete %s %s: %s", resourceName, id, response.getBody()),
-      response.getStatusCode(), is(204));
+      response.getStatusCode(),
+      is(204));
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/rest/support/http/ResourceClient.java]: Why has this line been split?

File: src/test/java/org/folio/rest/api/TestBaseWithInventoryUtil.java (hunk 9 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/rest/api/TestBaseWithInventoryUtil.java]: I've moved data provider classes to a separate package, and `src/test/java/org/folio/rest/api/testdata/ItemEffectiveLocationTestDataProvider.java` uses this properties.

File: src/test/java/org/folio/rest/support/http/ResourceClient.java (hunk 10 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/rest/support/http/ResourceClient.java]: For better readability. Now each param is on a separate line, i guess it is more readable now.

Comment (marcjohnson-kint) [reply src/test/java/org/folio/rest/support/http/ResourceClient.java]: Why does having the property and the matcher on separate lines make this more readable?

Comment (bohdan-suprun) [reply src/test/java/org/folio/rest/support/http/ResourceClient.java]: at least you can easily see which parameters are passed to the method. Previously it was like the getStatusCode() is a parameter of String.format(...).

Comment (marcjohnson-kint) [reply src/test/java/org/folio/rest/support/http/ResourceClient.java]: I really don't follow. 

The only changes reported here show splitting the `response.getStatusCode()` and `is(204)` onto separate lines, and I don't understand why that is more readable. 

What am I missing?

Comment (bohdan-suprun) [reply src/test/java/org/folio/rest/support/http/ResourceClient.java]: I've reverted this change.

**Comment by marcjohnson-kint** [comment]
@julianladisch are you content with the changes that @bohdan-suprun has made to the tests?

**Comment by marcjohnson-kint** [comment]
Thanks @julianladisch
